kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver_with_constraints.h

```diff
@@ -0,0 +1,1522 @@
+     * @param pScheme The integration scheme considered
+     * @param rModelPart The model part of the problem to solve
+     */
+    void SetUpDofSetWithConstraints(
+        typename TSchemeType::Pointer pScheme,
+        ModelPart& rModelPart
+        )
+    {
+        KRATOS_TRY;
+
+        KRATOS_INFO_IF("ResidualBasedEliminationBuilderAndSolverWithConstraints", ( this->GetEchoLevel() > 1 && rModelPart.GetCommunicator().MyPID() == 0)) << "Setting up the dofs" << std::endl;
+
+        DofsVectorType dof_list, second_dof_list; // NOTE: The second dof list is only used on constraints to include master/slave relations
+
+        typedef std::unordered_set < DofPointerType, DofPointerHasher> set_type;
+
+        // Declaring temporal variables
+        DofsArrayType dof_temp_all, dof_temp_solvable, dof_temp_slave;
+
+        // We assign an empty dof array to our dof sets
+        BaseType::mDofSet = DofsArrayType(); /// This corresponds with all the DoF of the system
+        mDoFSlaveSet = DofsArrayType();    /// This corresponds with the slave (the ones not solved after compacting the system using MPC)
+
+        /**
+         * Here we declare three sets.
+         * - The global set: Contains all the DoF of the system
+         * - The slave set: The DoF that are not going to be solved, due to MPC formulation
+         */
+        set_type dof_global_set, dof_global_slave_set;
+
+        #pragma omp parallel firstprivate(dof_list, second_dof_list)
+        {
+            ProcessInfo& r_current_process_info = rModelPart.GetProcessInfo();
+
+            // We cleate the temporal set and we reserve some space on them
+            set_type dofs_tmp_set, dof_temp_slave_set;
+            dofs_tmp_set.reserve(20000);
+            dof_temp_slave_set.reserve(200);
+
+            // Gets the array of elements from the modeler
+            ElementsArrayType& r_elements_array = rModelPart.Elements();
+            const int number_of_elements = static_cast<int>(r_elements_array.size());
+            #pragma omp for schedule(guided, 512) nowait
+            for (int i = 0; i < number_of_elements; ++i) {
+                auto it_elem = r_elements_array.begin() + i;
+
+                // Gets list of Dof involved on every element
+                pScheme->GetElementalDofList(*(it_elem.base()), dof_list, r_current_process_info);
+                dofs_tmp_set.insert(dof_list.begin(), dof_list.end());
+            }
+
+            // Gets the array of conditions from the modeler
+            ConditionsArrayType& r_conditions_array = rModelPart.Conditions();
+            const int number_of_conditions = static_cast<int>(r_conditions_array.size());
+            #pragma omp for  schedule(guided, 512) nowait
+            for (int i = 0; i < number_of_conditions; ++i) {
+                auto it_cond = r_conditions_array.begin() + i;
```

Comment:
No objection as such .. but just saying,  wouldn't 

```c++
const  auto begin_cond = r_conditions_array.begin()
#pragma omp for  schedule(guided, 512) nowait
            for (int i = 0; i < number_of_conditions; ++i) {
                auto it_cond = begin_cond + i;
```
be more efficient ?!

Reply:
Do you mean to move the begin outside?

Reply:
yes .. that way begin() is called only once !

Reply:
Ok